File: leetcode/Permutation_Sequence.java
import java.util.*;

public class Solution {
    int find(int pos, boolean[] has, int[] base, int[] fact, int n, int k) {
        int lim = fact[pos];
        int x = 0;
        for (int i = 1; i <= n; ++i)
            if (has[i]) {
                if (k <= lim) {
                    k -= lim - fact[pos];
                    x = i;
                    has[i] = false;
                    break;
                } else {
                    lim += fact[pos];
                }
            }
        if (pos == 0) return x;
        return x * base[pos] + find(pos - 1, has, base, fact, n, k);
    }
    public String getPermutation(int n, int k) {
        boolean[] has = new boolean[n + 1];
        for (int i = 1; i <= n; ++i) has[i] = true;
        int[] base = new int[n];
        base[0] = 1;
        for (int i = 1; i < n; ++i) base[i] = base[i - 1] * 10;
        int[] fact = new int[n];
        fact[0] = 1;
        for (int i = 1; i < n; ++i) fact[i] = i * fact[i - 1];

        int res = find(n - 1, has, base, fact, n, k);
        return Integer.toString(res);
    }
}

File: leetcode/Permutation_Sequence.cpp
class Solution {
    string generatePermutation(int n, int p, int k, 
                               vector<int> &num, vector<bool> &has) {
        if (p > n)
            return "";
        int x = 0;
        while (k > num[n - p] * (x + 1))
            ++x;
        k -= num[n - p] * x;
        for (int i = 1; i <= n; ++i)
            if (has[i]) {
                if (x == 0) {
                    has[i] = false;
                    return char('0' + i) + 
                           generatePermutation(n, p + 1, k, num, has);
                } else {
                    --x;
                }
            }
    }
public:
    string getPermutation(int n, int k) {
        vector<int> num(9);
        num[0] = 1;
        for (int i = 1; i <= 8; ++i)
            num[i] = num[i - 1] * i;

        vector<bool> has(10, true);

        return generatePermutation(n, 1, k, num, has);
    }
};
